### ISSUE TITLE
OSCAR protocol bug

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8338]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#466|
|Type|defect|
|Created|2004-01-05 20:32:16Z|

Attachments:

* [foo.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/cf2/cf2f328d24859d56d55d2b610b12525e60b21895/0f5843fe4a2bd32647fb23a438d1e43b1342c4aa.py) (435 bytes) - added by exarkun on 2004-01-05 20:32:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__466 466
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__words words
time__1073334736000000 1073334736000000
changetime__1470719198285738 1470719198285738
version__ 
owner__ 
cc__exarkun cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
The simplest OSCAR client I can think of fails with an exception from struct.unpack:

--- &lt;exception caught here> ---
  File "/home/exarkun/projects/python/Twisted/twisted/internet/default.py", line
527, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/home/exarkun/projects/python/Twisted/twisted/internet/tcp.py", line
250, in doRead
    return self.protocol.dataReceived(data)
  File "/home/exarkun/projects/python/Twisted/twisted/protocols/oscar.py", line
290, in dataReceived
    state=func(flap)
  File "/home/exarkun/projects/python/Twisted/twisted/protocols/oscar.py", line
373, in oscar_Data
    func(snac[2:])
  File "/home/exarkun/projects/python/Twisted/twisted/protocols/oscar.py", line
641, in oscar_0B_02
    self.reportingInterval = struct.unpack('!H',snac[3])[0]

Program to produce this is attached.  Assign correct values for an existing user
to "USER" and "PASS" before running.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="tdc's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tdc commented|
|-|-|

```
#!html
<pre>
i discussed this some time ago with z3p (after submitting same issue #333).
Solution (at least for me) was:

--- oscar.py.old 2003-10-30 19:02:18.000000000 +0100
+++ oscar.py  2003-11-11 18:28:24.000000000 +0100
@@ -638,7 +644,7 @@ class BOSConnection(SNACBased):
         """
         stats reporting interval
         """
-        self.reportingInterval = struct.unpack('!H',snac[3])[0]
+        self.reportingInterval = struct.unpack('!HH',snac[3])[0]
  
     def oscar_13_03(self, snac):
         """

_tdc_
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|z3p commented|
|-|-|

```
#!html
<pre>
Try oscar.py:1.19.  I committed a fix which should eliminate this problem.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

Example program no longer tracebacks.  I suppose this is resolved.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub commented|
|-|-|

In changeset 0c5a88ab36000416a057684e00f9c89b5ba3a5d4
```
#!CommitTicketReference repository="" revision="0c5a88ab36000416a057684e00f9c89b5ba3a5d4"
Merge pull request #466 from twisted/8727-rodrigc-ziploading-py3

Author: rodrigc
Reviewer: markrwilliams
Fixes: #8727

Allow twisted.trial.test.test_loader.ZipLoading tests to run on Python 3
```
